spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/actions/RewritePositionDeleteSparkAction.java

```diff
@@ -0,0 +1,456 @@
+          REWRITE_JOB_ORDER);
+  private static final Result EMPTY_RESULT =
+      ImmutableRewritePositionDeleteFiles.Result.builder().build();
+
+  private final Table table;
+  private final SparkBinPackPositionDeletesRewriter rewriter;
+
+  private int maxConcurrentFileGroupRewrites;
+  private int maxCommits;
+  private boolean partialProgressEnabled;
+  private RewriteJobOrder rewriteJobOrder;
+
+  RewritePositionDeleteSparkAction(SparkSession spark, Table table) {
+    super(spark);
+    this.table = table;
+    this.rewriter = new SparkBinPackPositionDeletesRewriter(spark(), table);
+  }
+
+  @Override
+  protected RewritePositionDeleteSparkAction self() {
+    return this;
+  }
+
+  @Override
+  public RewritePositionDeleteSparkAction filter(Expression expression) {
+    throw new UnsupportedOperationException("Regular filters not supported yet.");
+  }
+
+  @Override
+  public RewritePositionDeleteFiles.Result execute() {
+    if (table.currentSnapshot() == null) {
+      LOG.info("Nothing found to rewrite in empty table {}", table.name());
+      return EMPTY_RESULT;
+    }
+
+    validateAndInitOptions();
+
+    Map<StructLike, List<List<PositionDeletesScanTask>>> fileGroupsByPartition = planFileGroups();
+    RewriteExecutionContext ctx = new RewriteExecutionContext(fileGroupsByPartition);
+
+    if (ctx.totalGroupCount() == 0) {
+      LOG.info("Nothing found to rewrite in {}", table.name());
+      return EMPTY_RESULT;
+    }
+
+    Stream<RewritePositionDeletesGroup> groupStream = toGroupStream(ctx, fileGroupsByPartition);
+
+    RewritePositionDeletesCommitManager commitManager = commitManager();
+    if (partialProgressEnabled) {
+      return doExecuteWithPartialProgress(ctx, groupStream, commitManager);
+    } else {
+      return doExecute(ctx, groupStream, commitManager);
+    }
+  }
+
+  private Map<StructLike, List<List<PositionDeletesScanTask>>> planFileGroups() {
+    Table deletesTable =
+        MetadataTableUtils.createMetadataTableInstance(table, MetadataTableType.POSITION_DELETES);
+    CloseableIterable<PositionDeletesScanTask> scanTasks =
+        CloseableIterable.transform(
+            deletesTable.newBatchScan().ignoreResiduals().planFiles(),
+            t -> (PositionDeletesScanTask) t);
+
+    try {
+      StructType partitionType = Partitioning.partitionType(table);
+      StructLikeMap<List<PositionDeletesScanTask>> filesByPartition =
+          StructLikeMap.create(partitionType);
+
+      for (PositionDeletesScanTask task : scanTasks) {
+        StructLike coerced = coercePartition(task, partitionType);
+
+        List<PositionDeletesScanTask> partitionTasks = filesByPartition.get(coerced);
+        if (partitionTasks == null) {
+          partitionTasks = Lists.newArrayList();
+        }
+        partitionTasks.add(task);
+        filesByPartition.put(coerced, partitionTasks);
+      }
+
+      StructLikeMap<List<List<PositionDeletesScanTask>>> fileGroupsByPartition =
```

Comment:
for future: Can we explore the idea of having 2 helper methods: one for computing files by partition and another file groups by partition. Not in this PR.